src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/step/snapshot/AttemptSnapshotStep.kt

```diff
@@ -0,0 +1,75 @@
+package com.amazon.opendistroforelasticsearch.indexstatemanagement.step.snapshot
+
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.elasticapi.suspendUntil
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.model.ManagedIndexMetaData
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.model.action.SnapshotActionConfig
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.model.managedindexmetadata.StepMetaData
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.step.Step
+import org.apache.logging.log4j.LogManager
+import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotRequest
+import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotResponse
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.service.ClusterService
+import org.elasticsearch.rest.RestStatus
+
+class AttemptSnapshotStep(
+    val clusterService: ClusterService,
+    val client: Client,
+    val config: SnapshotActionConfig,
+    managedIndexMetaData: ManagedIndexMetaData
+) : Step(name, managedIndexMetaData) {
+
+    private val logger = LogManager.getLogger(javaClass)
+    private var stepStatus = StepStatus.STARTING
+    private var info: Map<String, Any>? = null
+
+    @Suppress("TooGenericExceptionCaught")
+    override suspend fun execute() {
+        try {
+            logger.info("Executing snapshot on ${managedIndexMetaData.index}")
+            val createSnapshotRequest = CreateSnapshotRequest()
+                .indices(managedIndexMetaData.index)
+                .snapshot(config.snapshot)
+                .repository(config.repository)
+            if (config.includeGlobalState != null) {
+                createSnapshotRequest.includeGlobalState(config.includeGlobalState)
+            }
+
+            val response: CreateSnapshotResponse = client.admin().cluster().suspendUntil { createSnapshot(createSnapshotRequest, it) }
+            when (response.status()) {
+                RestStatus.ACCEPTED -> {
+                    stepStatus = StepStatus.COMPLETED
+                    info = mapOf("message" to "Creating snapshot in progress for index: ${managedIndexMetaData.index}")
+                }
+                RestStatus.OK -> {
+                    stepStatus = StepStatus.COMPLETED
+                    info = mapOf("message" to "Snapshot created for index: ${managedIndexMetaData.index}")
+                }
+                else -> {
+                    stepStatus = StepStatus.FAILED
+                    info = mapOf("message" to "There was an error during snapshot creation for index: ${managedIndexMetaData.index}")
+                }
+            }
+        } catch (e: Exception) {
```

Comment:
Do failures such as no repository exists, duplicate snapshot name, etc. throw an exception and get caught here or do they return successfully with RestStatus.INTERNAL_SERVER_ERROR? If it's the latter, we probably still want to capture the relevant error message and add it to the info object above.

Reply:
Done